### ISSUE TITLE
Does FindFirstFile() work on all versions of Windows?

### ISSUE THREAD

**Issue by dancrepeau** [msel-source/meflib#20]
The FindFirstFile() API function doesn't seem to always work (meflib.c, line 4019).  For me it often returns error code of 3, which is ERROR_PATH_NOT_FOUND.  I am running Windows 7 SP1 - in the thread below the original poster states this API function does not work in Windows 7 SP1 - although I can't verify if that claim is true.

https://social.msdn.microsoft.com/Forums/SECURITY/en-US/f06d84cb-6f8d-4b69-bedb-f317df06bb6a/win32-apis-findfirstfile-findnextfile-findfirstfileex-do-not-work-on-windows-7-x64-sp1-retail?forum=vcgeneral

**Comment by dancrepeau** [comment]
It seems it does work, but the first parameter (temp_path) to FindFirstFile needs to be a "wide string".  Converting the string to a wide string using mbstowcs(), then caling FindFirstFile worked.  Then after calling FindNextFile, the results (in fdFile.cFileName) then need to be converted back to a normal string (using wcstombs()).

Another thing I found is sometimes temp versions of segments are found, so if you do a dir on a diretory, you might see (".", "..", and "AVF.segd") but there might also be "._AVF.segd", which should probably be ignored.  This can be done by changing the strcmp((si1*)temp_result, ".") to strncmp((si1*)temp_result, ".", 1).  This only compares the first character, which would catch all of the cases we care about: ".", "..", and "._whatever".

**Comment by cimbi** [comment]
is this connected to #19 ? I cannot help with windows much, I don't have a machine to test this on. What about @MaxvandenBoom ? Have you seen something like this happen?

**Comment by dancrepeau** [comment]
I don't think it's connected to the other topic, but I ran into this while writing a testbed for the other topic.  

Anyways, attached is the modified generate_file_list() (for _WIN32) that worked for me.  I am using Win 7 SP 1 OS and Visual Studio Pro 2017.  If other people run into this same problem, then this code should be considered as a replacement for what's in meflib.c currently.

[snippet.c.zip](https://github.com/msel-source/meflib/files/4893386/snippet.c.zip)

**Comment by MaxvandenBoom** [comment]
> is this connected to #19 ? I cannot help with windows much, I don't have a machine to test this on. What about @MaxvandenBoom ? Have you seen something like this happen?

@cimbi I haven't encountered this problem (Win 7 SP1 64-bit; Visual Studio 2015; Python and Matlab are both 64-bit)
@dancrepeau Have you been able to reproduce the problem using a specific path? It is hard to imagine it occurs at random. Just let it print the path on the error, then, given a problematic path, this should be very easy to fix.

Some possible causes could be a too long path-length (Windows is annoying like that), network paths, or running it in a 32-bit environment (from VS or a 32-bit Python; see Wow64DisableWow64FsRedirection)

Nice that you already have a code snippet that survives these problematic paths!! However, it might also be worth preventing an erroneous path string from ending up in the `FindFirstFile` function in the first place, if that indeed happens..
And if `FindFirstFile `indeed expects wide strings, then it would be great to implement that code snippet as well. More robustness, more better :)

Let me know if I can help

**Comment by dancrepeau** [comment]
All good ideas.  For my test program I am hard-coding a path in:

sprintf(in_file_name, "c:/test_data/I71.mefd/AVF.timd");

Using breakpoints I can see that no corruption is happening on that string, and that the path itself is completely valid and I have permissions to read it. However, the FindFirstFile simply doesn't see the path.  (But it does see the path after I convert the same string to the wide string).  If FindFirstFile really does require wide strings, then it seems crazy to me that no other user has ever run into this problem!  So I do wonder if it's a Windows 7 issue.  If it's not harmful on other versions of Windows, then perhaps the changes should be adopted universally - although I am hesitant to push an update that could potentially break other implementations.

**Comment by MaxvandenBoom** [comment]
I can give that path a try later this evening.

What happens if you add a `/` to that path? (`.timd` is a folder right?)
The function might assume a file now due to the missing directory separator character.

**Comment by vkremen** [comment]
I also realized that it could might have problems when there is dash in the name e.g "E1-E2". But didn't really have time to test it thoroughly.

Would you be able to test it as well.

Thanks,


- Vaclav

 ​

Vaclav Kremen, M.Sc.Eng., Ph.D.

________________________________________________________________

Associate Professor of Biomedical Engineering, Principal Engineer

@ Bioelectronics Neurophysiology and Engineering Lab  <https://www.mayo.edu/research/labs/bioelectronics-neurophysiology-engineering/overview>   |  Dept. of Neurology

Phone: (507) 538-5698  |  Pager: 127-09723  |  Email: kremen.vaclav@mayo.edu<mailto:kremen.vaclav@mayo.edu>

ORCID: https://orcid.org/0000-0001-9844-7617

________________________________________________________________

Mayo Clinic  |  200 First Street SW, Rochester, MN 55905  |  www.mayoclinic.org<http://www.mayoclinic.org/>


________________________________

**Comment by dancrepeau** [comment]
The way the path is specified is something like: "c:\\test_data\\I71.mefd\\AVF.timd\\*.*".  The backslashes are a single character (ascii 92).  So when looking for files within that, the *.* mask should find the .segd folders.  I tired using just a sing * rather than *.*, but that made no difference.

I doubt a dash would make any difference, as the segment folders all have a dash in them already.  (I know you are referring to dashes in the channel name).  That's something that could be tested separately.

**Comment by vkremen** [comment]
sounds good and reasonable to me. Thanks, Dan.


- Vaclav

 ​

Vaclav Kremen, M.Sc.Eng., Ph.D.

________________________________________________________________

Associate Professor of Biomedical Engineering, Principal Engineer

@ Bioelectronics Neurophysiology and Engineering Lab  <https://www.mayo.edu/research/labs/bioelectronics-neurophysiology-engineering/overview>   |  Dept. of Neurology

Phone: (507) 538-5698  |  Pager: 127-09723  |  Email: kremen.vaclav@mayo.edu<mailto:kremen.vaclav@mayo.edu>

ORCID: https://orcid.org/0000-0001-9844-7617

________________________________________________________________

Mayo Clinic  |  200 First Street SW, Rochester, MN 55905  |  www.mayoclinic.org<http://www.mayoclinic.org/>


________________________________

**Comment by MaxvandenBoom** [comment]
> The way the path is specified is something like: "c:\test_data\I71.mefd\AVF.timd\_._".

If you want to test, we should be sure what exactly goes into the FindFirstFile function ("something like" might not be enough). Can you provide the exact string that goes into the FindFirstFile function when it goes wrong? That would help greatly..

Also I think a dash should not be a problem, but it might be good to test.

**Comment by dancrepeau** [comment]
Sure, sorry, didn't mean to be vague.  Attached is a screenshot of the exact string (as seen in the debugger), which works under my modified code, but not under the code currently in meflib.c

![test_path](https://user-images.githubusercontent.com/42556139/87075841-ce350f00-c1d5-11ea-9fec-c0298d7cdca9.jpg)

**Comment by MaxvandenBoom** [comment]
Cool, thanks! I will test it later and let you guys know.

**Comment by MaxvandenBoom** [comment]
Hi Dan,

I tried it with this simple sample program:


```
#include <windows.h>

int main() {

	char path[1024];
	sprintf(path, "%s", "C:\\test_data\\I71.mefd\\AVF.timd\\*.*");
	printf("path: %s\n", path);
	
	WIN32_FIND_DATA fdFile;
	HANDLE hFind = FindFirstFile(path, &fdFile);
	if (hFind == INVALID_HANDLE_VALUE) {
		printf("Error - INVALID_HANDLE_VALUE\n");

	} else if (hFind == ERROR_PATH_NOT_FOUND) {
		printf("Error - ERROR_PATH_NOT_FOUND\n");

	} else {
		printf("Path found\n");

	}
	FindClose(hFind);

	getchar();

    return 0;
}
```

and it just always seems to work. I tried an empty directory and I tried putting a segment in the directory, both fine.
Maybe I am doing something different? This is what should go wrong, right?
Perhaps you could try to run the program above and see if you get an error?

**Comment by dancrepeau** [comment]
Hi Max - I created a console project in Visual Studio 2017 (still using Win 7 SP 1 OS), and the only change I made to your test code is to add #include <stdio.h>, then turned off the precompiled header option, and turned off the warning for sprintf being insecure.  Same results as I've previously gotten...

![maxtest](https://user-images.githubusercontent.com/42556139/87201709-fd21b280-c2b3-11ea-8319-59e1990bfcc4.jpg)

**Comment by dancrepeau** [comment]
With the changes added to make it a wide string - works fine.

![maxtest2](https://user-images.githubusercontent.com/42556139/87202379-47effa00-c2b5-11ea-8a51-62c2300cb557.jpg)

**Comment by dancrepeau** [comment]
(and "c:" or "C:" at the beginning of the path makes no difference - I tried both tests both ways).

**Comment by MaxvandenBoom** [comment]
That is pretty absurd :)

Do you have anything in the folder? Did you try it with the folder being empty?
At least there is a solution, I'm just intrigued by why it would work differently.
Which filesystem do you have for c:, NTFS?

**Comment by dancrepeau** [comment]
Agreed, it's absurd.  :)  Yes, the folder contains "AVF-000000.segd" which contains real data, and I can read that data after I substitute the modified generate_file_list().  It is NTFS for my c drive.  I have no other ideas about the cause - in my original post there was a stack exchange thread where the OP thought it was the OS (which happens to be the same I am using).  It seems unlikely... but I have no better theories.

**Comment by MaxvandenBoom** [comment]
Weird...
I guess it should be good to use the wide-char solution anyway, so mef will be more compatible with tropical characters in filepaths. The generate_file_list is already in a windows precompiler statement anyway.

Just out of curiousity:
- Could you try it with the directory being empty?
- If you replace `FindFirstFile `with `FindFirstFileA`, does it work?
- Do you have a Locale set on your computer?

**Comment by dancrepeau** [comment]
I just upgraded to Windows 10 (enterprise), and Visual Studio 2019.  FindFirstFile() still didn't work, as before.

1) Directory being empty (meaning directory is there but contains nothing) has no effect on the result.
2) FindFirstFileA() works - that seems to solve the problem.
3) Not sure about Locale, I don't (by default) have admin privileges, I could check this if it's important, but I don't think we can assume users will have admin privileges on their machines.

**Comment by MaxvandenBoom** [comment]
> 2. FindFirstFileA() works - that seems to solve the problem.

Ah, that is interesting. I think I have an idea. If you move the cursor over `FindFirstFile`, right-click and select 'Go to definition', where do you end up?

I end up in fileapi.h with this line:
`#define FindFirstFile  FindFirstFileA`
which basically makes `FindFirstFile` an alias for `FindFirstFileA`. My guess is that you end up with a different definition.

Is UNICODE set anywhere in your project's preprocessor definitions? Or a character set specified?

**Comment by dancrepeau** [comment]
Correct, I get:
#define FindFirstFile FindFirstFileW
So... going back to generate_file_list(), if I add an 'A' to the end of FindFirstFile(), FindNextFile(), as well as WIN32_FIND_DATA (the struct itself), then that solves the problem for me.  Although I'll note there is still the bug where it finds duplicates, since for me each segment is showing up with a duplicate with a '.' in front of the name.

Yes, my project has _UNICODE defined.  This is the default when I created the solution (C++ Console app).  Perhaps non-c++ projects have different defaults.

**Comment by MaxvandenBoom** [comment]
Mystery solved :) So it depends on the project/compiler settings whether it uses `Find...FileA` or `Find....FileW`; and therefore whether Unicode's special characters will work in the filepath?

For windows, it might be good to enforce the `W` versions of these kind of functions,. The generate_file_list already has a windows precompiler block, so it should not interfere with the linux/mac workings. What do you think?

**Comment by MaxvandenBoom** [comment]
I actually tested it really quickly and `Find..FileA` works perfectly with special characters in the filepaths as well (at least the latin-set ones: `á` `ë` etc).

I am unsure whether there is a big benefit in using `Find..FileW` and wide-chars in the meflib situation, chinese characters? Longer (>260 char) filenames?

Enforcing `Find...FileA` might be a much more straightforwards solution... It might save you the headache of taking `wchars` into consideration for all the surrounding code as well, not to mention the windows-specific problems that might come it.

**Comment by dancrepeau** [comment]
Completely agree, I think FileA is the way to go... just force it to be that way, to remove ambiguity.  It seems my initial issue was selecting a C++ project, as opposed to C# or some other project type, which gave me a different set of precompiled settings relative to what other people are using.

**Comment by vkremen** [comment]
I think in many situations, one might have quite long file names if including file path. Is the file path included in this discussion in filename as well. If yes, I would like to suggest to go for >260 chars. If it is just the filename itself, <260 chars is probably fine.


- Vaclav

 ​

Vaclav Kremen, M.Sc.Eng., Ph.D.

________________________________________________________________

Associate Professor of Biomedical Engineering, Principal Engineer

@ Bioelectronics Neurophysiology and Engineering Lab  <https://www.mayo.edu/research/labs/bioelectronics-neurophysiology-engineering/overview>   |  Dept. of Neurology

Phone: (507) 538-5698  |  Pager: 127-09723  |  Email: kremen.vaclav@mayo.edu<mailto:kremen.vaclav@mayo.edu>

ORCID: https://orcid.org/0000-0001-9844-7617

________________________________________________________________

Mayo Clinic  |  200 First Street SW, Rochester, MN 55905  |  www.mayoclinic.org<http://www.mayoclinic.org/>


________________________________

**Comment by dancrepeau** [comment]
The filename itself is 256 chars (in bytes... possibly fewer chars in unicode) and the whole pathname including the filename is 1024 chars.

**Comment by vkremen** [comment]
I see. O.k. That should be fine, then, I guess.


- Vaclav

 ​

________________________________

**Comment by MaxvandenBoom** [comment]
> I think in many situations, one might have quite long file names if including file path. Is the file path included in this discussion in filename as well. If yes, I would like to suggest to go for >260 chars. If it is just the filename itself, <260 chars is probably fine.

I agree with Vaclav on the length, but I might have been mislead. After reading a bit more I don't think `Find...FileW` will allow for longer filepaths over `Find...FileA`, mea culpa. From what I understand, the only advantage would be support for an even more wide character-set, unlikely to be encountered.

**Comment by dancrepeau** [comment]
It seems to me we can just pick one (unicode or non-unicode) and go with it until a need arises to change it, but I think the ambiguity of not specifying Find...FileA vs Find...FileW is very problematic in that code becomes depending on having the correct precompiler definitions.

**Comment by MaxvandenBoom** [comment]
Definitely true. From what I understand if you are aiming for multi-OS (also taking into consideration the surrounding routines) it is a bad idea to mix char/wchar and string/wstring use

**Comment by cimbi** [comment]
Has this been resolved? Can we close the issue?

**Comment by MaxvandenBoom** [comment]
> Has this been resolved? Can we close the issue?

Not yet, but the solution is very simple. I would recommend just changing the `Find...File` calls to `Find...FileA` calls.
But I don't know if Dan is still on the fence on using A or W..

**Comment by dancrepeau** [comment]
Yes, I am happy with the 'A' solution.  I will update the code soon with the changes.

**Comment by cimbi** [comment]
Did the commit 579afaf1b3b870b5bce4b904f9ceb3f4f514b6b0 fix this issue? What about issues #19 and #18?

**Comment by dancrepeau** [comment]
Correct, this issue is resolved, as well as issue #18.

**Comment by dancrepeau** [comment]
Correct, this issue is resolved, as well as issue #18.
